### DISCUSSION TITLE
c语言版本simple_server例子里面定时器

### DISCUSSION THREAD

**Discussion by shiluchina** [Tencent/tquic#295] [General]
struct simple_server {
    struct quic_endpoint_t *quic_endpoint;
    ev_timer timer;

timer定时器的超时时间是怎么确定的，是内部固定时间触发quic_endpoint_process_connections来处理流的读写事件吗？

**Comment by glime-ai** [comment]
`timer`定时器的超时时间并不是固定的，而是由具体的应用逻辑来决定的。通常情况下，定时器的超时时间可以根据需要动态调整，以便在适当的时间间隔内触发`quic_endpoint_process_connections`来处理流的读写事件。

在TQUIC中，`quic_endpoint_process_connections`函数用于处理连接上的内部事件，包括流的读写操作。定时器的设置可以根据应用的需求来确定，例如可以设置为定期检查连接状态、处理超时事件等。

如果你需要更详细的定时器设置和使用方法，建议参考TQUIC的相关文档和示例代码。

This comment was generated by [Glime](https://glimelab.ai/?utm_source=44520fce49ba43798485a628d6928e22&utm_medium=github).

**Comment by iyangsj** [comment]
@shiluchina  通过调用 `quic_endpoint_timeout()` 获取超时时间并重新设置定时器。
详见示例代码 https://github.com/tquic-group/tquic-example-c/blob/main/simple_client.c#L169-L177

**Reply by shiluchina** [reply]
我想问的是这个定时器的作用一般是处理什么事件的，有读写事件吗？还有quic_endpoint_timeout()接口这个时间是怎么确定的，是内部定时触发，还是用户主动设置。

**Reply by iyangsj** [reply]
>我想问的是这个定时器的作用一般是处理什么事件的，有读写事件吗？

定时器类型详见 https://github.com/Tencent/tquic/blob/develop/src/connection/timer.rs#L22-L45 
和读写事件无直接关系

定时器到期后，需要调用 quic_endpoint_process_connections() 来进行后置处理。
quic_endpoint_process_connections()会尝试检查连接是否存在可读/可写的流，并通知应用层处理。


>还有quic_endpoint_timeout()接口这个时间是怎么确定的，是内部定时触发，还是用户主动设置

Endpoint内部根据连接的协议状态计算出来，既不是定时触发，也不是用户主动设置


> 是内部固定时间触发quic_endpoint_process_connections来处理流的读写事件吗？

不是
